src/vaev-engine/style/rules.cpp

```diff
@@ -257,4 +257,296 @@ void MediaRule::repr(Io::Emit& e) const {
+                        nonLookupRules.pushBack({ruleId, rule});
+                    } else {
+                        ruleIdToNeededCount.put(ruleId, conditionsCount);
+                    }
+                } else if (s.type == Nfix::OR) {
+                    // NOTE: If an element has 2 or more occourence of this rule in its list, we can assume
+                    // the rule as matched, since at least one of the occourences is due to a lookupable selector,
+                    // which is guaranteed to match
+                    bool hasNonLookupable = false;
+                    for (auto const& inner : s.inners) {
+                        if (lookupIsMatch(inner)) {
+                            _buildRuleFromSelector(rule, ruleId, inner);
+                        } else {
+                            hasNonLookupable = true;
+                        }
+                    }
+                    if (hasNonLookupable)
+                        nonLookupRules.pushBack({ruleId, rule});
+                } else {
+                    nonLookupRules.pushBack({ruleId, rule});
+                }
+            },
+            [&](auto const&) {
+                nonLookupRules.pushBack({ruleId, rule});
+            },
+        });
+    }
+
+    void _buildRule(Cursor<Rule> rule, Media media) {
+        rule->visit(Visitor{
+            [&](StyleRule const& r) {
+                _ruleCount++;
+                _buildRuleFromSelector(&r, _ruleCount, r.selector);
+            },
+            [&](MediaRule const& r) {
+                if (r.match(media))
+                    for (auto const& subRule : r.rules)
+                        _buildRule(&subRule, media);
+            },
+            [&](auto const&) {
+                // Ignore other rule types
+            },
+        });
+    }
+
+    void build(Media media, StyleSheetList const& styleBook);
+
+    static bool lookupIsMatch(AttributeSelector const& selector) {
+        if (selector.match != AttributeSelector::Match::PRESENT and
+            selector.match != AttributeSelector::Match::EXACT)
+            return false;
+
+        // NOTE: Attr selectors with non-qualified namespaces could also be lookupable selectors, but currently
+        // they are not due to increased code complexity.
+        return selector.qualifiedName.ns != NONE;
+    }
+
+    static bool lookupIsMatch(TypeSelector const& selector) {
+        // NOTE: All type selectors could be lookupable selectors, but currently not all of them are due to increased
+        // code complexity.
+        return selector.qualifiedName.ns == NONE &&
+               selector.qualifiedName.name != NONE;
+    }
+
+    static bool lookupIsMatch(Selector const& selector) {
+        if (auto s = selector.is<AttributeSelector>())
+            return lookupIsMatch(*s);
+
+        if (auto s = selector.is<TypeSelector>())
+            return lookupIsMatch(*s);
+
+        return selector.is<IdSelector>() || selector.is<ClassSelector>();
+    }
```

Comment:
The name of this function is too vague, also use `or` instead of `||`

Reply:
gotta learn those prompts